fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java

```diff
@@ -182,20 +190,98 @@ private void maybeLogBuildProgress(SubspaceProvider subspaceProvider, byte[] ret
+        validateOrThrowEx(srcIndex != null, "source index is null");
         // idempotence - We could have verified it at the first iteration only, but the repeating checks seem harmless
         validateOrThrowEx(maintainer.isIdempotent(), "target index is not idempotent");
+        // readability - This method shouldn't block if one has already opened the record store (as we did)
+        validateOrThrowEx(store.isIndexReadable(srcIndex), "source index is not readable");
 
-        // this should never happen. But it makes the compiler happy
-        validateOrThrowEx(srcIndex != null, "source index is null");
+        RangeSet rangeSet = new RangeSet(store.indexRangeSubspace(index));
+        AsyncIterator<Range> ranges = rangeSet.missingRanges(store.ensureContextActive(), startBytes, endBytes).iterator();
+
+        final ExecuteProperties.Builder executeProperties = ExecuteProperties.newBuilder()
+                .setIsolationLevel(IsolationLevel.SNAPSHOT)
+                .setReturnedRowLimit(getLimit()); // always respectLimit in this path
+        final ScanProperties scanProperties = new ScanProperties(executeProperties.build());
+
+        return ranges.onHasNext().thenCompose(hasNext -> {
+            if (Boolean.FALSE.equals(hasNext)) {
+                return AsyncUtil.READY_FALSE; // no more missing ranges - all done
+            }
+            Range range = ranges.next();
+            Tuple rangeStart = Arrays.equals(range.begin, START_BYTES) ? null : Tuple.fromBytes(range.begin);
+            Tuple rangeEnd = Arrays.equals(range.end, END_BYTES) ? null : Tuple.fromBytes(range.end);
+
+            RecordCursor<FDBIndexedRecord<Message>> cursor =
+                    store.scanIndexRecords(srcIndex, IndexScanType.BY_VALUE, TupleRange.between(rangeStart, rangeEnd), range.begin, scanProperties);
+
+            final AtomicLong recordsScannedCounter = new AtomicLong();
+            final AtomicReference<RecordCursorResult<FDBIndexedRecord<Message>>> lastResult = new AtomicReference<>(RecordCursorResult.exhausted());
+            final AtomicBoolean isEmpty = new AtomicBoolean(true);
+
+            return iterateRangeOnly(store, cursor,
+                    IndexingByIndex::castCursorResultToStoreRecord,
+                    lastResult::set,
+                    isEmpty, recordsScannedCounter
+            ).thenCompose(vignore -> {
+                long recordsScannedInTransaction = recordsScannedCounter.get();
+                recordsScanned.addAndGet(recordsScannedInTransaction);
+                if (common.isTrackProgress()) {
+                    store.context.ensureActive().mutate(MutationType.ADD, scannedRecordsSubspace.getKey(),
+                            FDBRecordStore.encodeRecordCount(recordsScannedInTransaction));
+                }
+                final byte[] realEnd =
+                        isEmpty.get() ? null : lastResult.get().getContinuation().toBytes();
```

Comment:
This feels like there's a lot of code dupe between this and the other `buildRangeOnly`.

I think some of that could be removed if the `iterateRangeOnly` took a lambda that would take a record and produced a tuple to use for the range set. For the `byRecords` method, it would return `getPrimaryKey()` but for this version, it would return `.getIndexEntry().getKey()`. And then `iterateRangeOnly` could update the range set (maybe it gets a name change), and then also, you might be able to have unified logic for selecting missing ranges, etc.

There's also the idea from Scott Dugas about packaging some of that up into a "records producer" that knows how to produce records within "some range" and also tell you the proper key to use for a given record.

Reply:
Great point! 
I did use a simplified logic in the `indexingByIndex`. If it passed all the tests and successfully used in a controlled environment, I was hopping to return and simplify the `indexingByRecords` in a similar way (and handle the rangeSet in this function) - making your code sharing suggestion much easier. Would you agree?